### PR TITLE
Add a second job that watches rekor for new STH entries and publishes

### PR DESCRIPTION
**Pull request by dlorenc**
them to GCS.

We don't always have a new one, and under load there might be "batches" of entries
all integrated into the same STH. This means there is no guaranteed frequency of updates
or even a guarantee that every index will exist.

The values (and timestamps) should be monotonically increasing though.

Signed-off-by: Dan Lorenc <dlorenc@google.com>

### CODE REVIEW

File: config/watcher.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      containers:
+      - name: rekor-watcher
+        image: ko://github.com/sigstore/rekor/cmd/server
+        command: ["/ko-app/server"]
+        ports:
+        - containerPort: 3000
+        - containerPort: 2112 # metrics
+        args: [
+          "watch",
+          "--rekor_server.address=rekor-server",
```
Comment (lukehinds) [inline config/watcher.yaml:30]: Does that not need to be rekor-watcher (the binding hostname)?

Comment (dlorenc) [reply config/watcher.yaml:30]: Nope! This is how the watcher finds the server. I reused the same flag for consistency, but could also see how that's more confusing...

File: cmd/server/app/watch.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+			if err := uploadToGcs(ctx, bucket, lr); err != nil {
+				log.Logger.Warnf("error uploading result: %s", err)
+				continue
+			}
+			last = lr
+		}
+	},
+}
+
+func init() {
```
Comment (font) [inline cmd/server/app/watch.go:134]: There's a duplicate `init` at the bottom of this file.

**Review by font — COMMENTED** [review]
@dlorenc this is cool! So now we can add others (likely as sub-commands to `watch`) to support https://github.com/sigstore/rekor/issues/191.

**Review by font — COMMENTED** [review]
@dlorenc Is there any way to add an e2e test so that when we implement others from https://github.com/sigstore/rekor/issues/191 we know we won't have broken GCS?

File: cmd/server/app/watch.go (hunk 2 above)
Comment (dlorenc) [reply cmd/server/app/watch.go:134]: Thanks!

**Comment by dlorenc** [comment]
> @dlorenc Is there any way to add an e2e test so that when we implement others from #191 we know we won't have broken GCS?

Yes, really good idea. I need to think through the right layer to mock out GCS at...

**Comment by dlorenc** [comment]
> Yes, really good idea. I need to think through the right layer to mock out GCS at...

All set!

File: tests/e2e_test.go [hunk 3]
```diff
@@ -283,3 +286,28 @@ func TestUploadNoAPIKeyInOutput(t *testing.T) {
 		t.Errorf("CLI output contained API key when it should have squelched it")
 	}
 }
+
+func TestWatch(t *testing.T) {
+
+	td := t.TempDir()
+	cmd := exec.Command("../rekor-server", "watch", "--interval=1s")
```
Comment (font) [inline tests/e2e_test.go]: `s|"../rekor-server"|server|`

File: cmd/server/app/watch.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		}
+
+		pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+		if err != nil {
+			return err
+		}
+
+		verifier := tclient.NewLogVerifier(hasher.DefaultHasher, pub, crypto.SHA256)
+		ctx := context.Background()
+		bucketURL := os.Getenv("REKOR_STH_BUCKET")
```
Comment (font) [inline cmd/server/app/watch.go]: The next function call will catch this, but we may want to consider fatally logging a specific error message if this env var is unset.

**Review by font — APPROVED** [review]
@dlorenc Nice!

File: cmd/server/app/watch.go (hunk 4 above)
Comment (dlorenc) [reply cmd/server/app/watch.go]: Done!

File: tests/e2e_test.go (hunk 3 above)
Comment (dlorenc) [reply tests/e2e_test.go]: I don't think so... We compile it as "rekor-server" above, and the test directory here is set to test/, so we need to go back up one.

Comment (font) [reply tests/e2e_test.go]: Oh sorry I should have been more specific! The string replacement was to use the `server` constant var defined below as I assumed that's what you intended and I didn't see the constant used anywhere else.

**Review by font — APPROVED** [review]

**Review by font — APPROVED** [review]
LGTM. Can we merge this?

**Pull request merged**